### PR TITLE
feat: add totp support in backend

### PR DESCRIPTION
**Pull request by Steve0x2a**
1. add /api/totp[get] to implement user request to set totp, and return secret and recovery-code.
2. add /api/totp[post] to implement user initial authentication of secret and code and enable 2fa feature.
3. add /api/totp[delete] to implement the user to use recovery-code to delete the totp secret that has been set.

However, no front-end support has been implemented yet, so this pr is currently limited to back-end calls, and the back-end code of totp applied to the current login process still needs to be improved.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@seriouszyx plz review

**Review by seriouszyx — APPROVED** [review]
LGTM for purely looking at the code and simply testing the APIs. It's better to provide front-end pages later.

**Comment by seriouszyx** [comment]
@Steve0x2a I create an option in the account edit page as below. Currently, when enabling the option, I call the init API, and when disabling the option,  I call the delete API. Is this correct? What `is2fa` is used for? When to use set API?

image

**Comment by Steve0x2a** [comment]
> @Steve0x2a I create an option in the account edit page as below. Currently, when enabling the option, I call the init API, and when disabling the option, I call the delete API. Is this correct? What `is2fa` is used for? When to use set API?
> 
> image

I am using the "is2fa" flag to confirm that the user is currently enabled for two-step authentication. When the user wants to open 2fa, request the init API, which will return the secret and recovery-code, then the user needs to generate a code in the local 2fa client, submit this code through the set API, and casdoor verifies that the code was generated using the previous secret, and finally accept this totp setting.

**Comment by hsluoyz** [comment]
@Steve0x2a @seriouszyx plz finish the frontend in this PR

**Comment by seriouszyx** [comment]
@hsluoyz Can I push my commit to this PR directly? Should I be assigned some permissions?

**Comment by seriouszyx** [comment]
@hsluoyz @Steve0x2a 
On the account page, an option is added to set 2-step verification. 

image

When enabling the option, it will jump to the `set-totp` page which is similar to forget page in style. Firstly, you should verify the password.

image

Secondly, the QRCode will be shown based on the URL generated by the backend. 

image

You can use a local 2fa client (e.g. Google Authenticator) to scan the QRCode. 

99aafc94f2325196804835224b9842f

It will generate a code.

742ac6979646096288a30cf2771a087

Submit the code and complete.

image

So the 2-step verification is enabled. 

image

When disabling the option, it will call the `delete-totp` API.

**Comment by Steve0x2a** [comment]
@seriouszyx nice job!
However, I think the recovery-code should only be shown to the user when initializing the totp, the subsequent steps of the api will not return the recovery-code. the user should enter the recovery-code when calling the delete api.

Of course, if sufficient security is not needed, the current solution is also feasible.

Our next step should be to verify the totp during login.

**Comment by hsluoyz** [comment]
@seriouszyx demo link?

**Comment by seriouszyx** [comment]
> @seriouszyx demo link?

http://82.156.216.21:8000/

**Comment by hsluoyz** [comment]
@Steve0x2a @seriouszyx plz fix:

image

**Comment by nodece** [comment]
Is it device flow authentication?

File: controllers/totp.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	if !ok {
+		return
+	}
+	user := object.GetUser(userId)
+	if user == nil {
+		c.ResponseError(fmt.Sprintf("The user: %s doesn't exist", userId))
+		return
+	}
+	if user.Is2fa {
+		c.ResponseError(fmt.Sprintf("The user: %s has turned on 2fa already", userId))
```
Comment (nodece) [inline controllers/totp.go:45]:
```suggestion
		c.ResponseError(fmt.Sprintf("The user: %s has two-factor authentication enabled", userId))
```

File: controllers/totp.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
```
Comment (nodece) [inline controllers/totp.go:81]: Same comment.

File: object/user.go [hunk 3]
```diff
@@ -59,6 +61,7 @@ type User struct {
 	IsGlobalAdmin     bool     `json:"isGlobalAdmin"`
 	IsForbidden       bool     `json:"isForbidden"`
 	IsDeleted         bool     `json:"isDeleted"`
+	Is2fa             bool     `json:"is2fa"`
```
Comment (nodece) [inline object/user.go:64]: Use `enableTwoFactor` looks clear.

File: routers/router.go [hunk 4]
```diff
@@ -151,6 +151,9 @@ func initAPI() {
 	beego.Router("/api/send-email", &controllers.ApiController{}, "POST:SendEmail")
 	beego.Router("/api/send-sms", &controllers.ApiController{}, "POST:SendSms")
 
+	beego.Router("api/totp", &controllers.ApiController{}, "GET:InitTOTP")
```
Comment (nodece) [inline routers/router.go:154]:
```suggestion
	beego.Router("api/get-totp", &controllers.ApiController{}, "GET:InitTOTP")
```

File: routers/router.go [hunk 5]
```diff
@@ -151,6 +151,9 @@ func initAPI() {
...
+	beego.Router("api/totp", &controllers.ApiController{}, "POST:SetTOTP")
```
Comment (nodece) [inline routers/router.go:155]:
```suggestion
	beego.Router("api/set-totp", &controllers.ApiController{}, "POST:SetTOTP")
```

File: controllers/totp.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import (
+	"fmt"
+
+	"github.com/casbin/casdoor/object"
+)
+
+type TOTPInit struct {
+	Secret        string `json:"secret"`
+	RecoveryCodes string `json:"recoveryCode"`
+	URL           string `json:"qrcodeUrl"`
```
Comment (nodece) [inline controllers/totp.go:25]:
```suggestion
	URL           string `json:"url"`
```

File: controllers/totp.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+// SetTOTP
+// @Title: SetTOTP
+// @Tag: TOTP API
+// @Description: Verify that the user has set the correct TOTP
+// @Param: secret  formData    string  true        "The totp secret returned by casdoor"
+// @Param: code  formData    string  true        "Code generated using totp secret"
+// @Success: controllers.Response The Response object
+// @router: /totp [post]
+func (c *ApiController) SetTOTP() {
```
Comment (nodece) [inline controllers/totp.go:70]:
```suggestion
func (c *ApiController) VerifyTOTP() {
```

File: object/user.go [hunk 8]
```diff
@@ -32,6 +32,8 @@ type User struct {
 	Type              string   `xorm:"varchar(100)" json:"type"`
 	Password          string   `xorm:"varchar(100)" json:"password"`
 	PasswordSalt      string   `xorm:"varchar(100)" json:"passwordSalt"`
+	TotpSecret        string   `xorm:"varchar(100)" json:"-"`
+	RecoveryCode      string   `xorm:"varchar(100)" json:"recoveryCode"`
```
Comment (nodece) [inline object/user.go:36]:
```suggestion
	TotpRecoveryCode      string   `xorm:"varchar(100)" json:"totpRecoveryCode"`
```

**Comment by hsluoyz** [comment]
@Steve0x2a plz fix:

image

**Pull request closed without merging**